### DISCUSSION TITLE
How to read file from embedded fs?

### DISCUSSION THREAD

**Discussion by Plorenzo** [pocketbase/pocketbase#3906] [Q&A] [answered]
Using Pocketbase as a framework, I want to assign a preexisting avatar to every new created user. I'm serving the UI as an embeded fs

```go
// Package ui handles the embedding of the static site
package ui

import (
	"embed"
	"github.com/labstack/echo/v5"
)

//go:embed all:build
var webDir embed.FS

// DistDirFS contains the embedded dist directory files (without the "dist" prefix)
var DistDirFS = echo.MustSubFS(webDir, "build")
```

my placeholder avatar file lives in `./build/img/user.png`

and this is how Im reading it, to save it in `OnModelAfterCreate` hook 

```go
app.OnModelAfterCreate().Add(func(e *core.ModelEvent) error {

		if e.Model.TableName() == "users" {
			// Adds default avatar image after user creation
			path := "img/user.png"

			file, err := ui.DistDirFS.Open(path)
			if err != nil {
				return err
			}
			defer file.Close()
			// Read the file into a byte slice
			fileBytes, err := io.ReadAll(file)
			if err != nil {
				return err
			}

			// Create a *filesystem.File from the byte slice
			embeddedFile, err := filesystem.NewFileFromBytes(fileBytes, "user.png")
			if err != nil {
				return err
			}

			form := forms.NewRecordUpsert(app, e.Model.(*models.Record))
			form.SetDao(daos.New(app.Dao().DB()))
			form.AddFiles("avatar", embeddedFile)

			if err := form.Submit(); err != nil {
				return err
			}

			return nil
		}

```

This works but it feels like a lot of work, is there another way? maybe to pass directly to `form.AddFiles()` the file form `DistDirFS.Open()`

Any suggestions welcome

**Comment by ganigeorgiev** [comment, marked as answer]
There is no other easier way at the moment.

If all users are expected to have the same initial avatar image, you can move the `filesystem.File` initialization outside of the hook to prevent creating it every time.

_Side-note: the call `form.SetDao(daos.New(app.Dao().DB()))` shouldn't be necessary (or if you want to use the hooks dao, you can pass `e.Dao` instead) and you can skip the if condition inside the hook by declaring the "users" in the `OnModelAfterCreate("users")` tags list._

**Comment by Plorenzo** [comment]
Noted, thanks for the tips. 

I have the `if` because I check also for other tables to do some other operations. Would it be recommended to use your suggestions instead? 

```go
app.OnModelAfterCreate("user").Add(func(e *core.ModelEvent) error {
// Do things
}

app.OnModelAfterCreate("other_table").Add(func(e *core.ModelEvent) error {
// Do things
}
// etc
```

VS

```go
app.OnModelAfterCreate().Add(func(e *core.ModelEvent) error {

		if e.Model.TableName() == "users" {
                //do things
               }
               if e.Model.TableName() == "other_table" {
                //do things
               }
}
```

**Reply by ganigeorgiev** [reply]
I think both approaches are fine.

Personally if you have a lot of code (eg. more than 50-100+ lines) in each condition block then I would probably split it into 2 hook calls or if there is a chance the same code to be used by more than one hook - move each if block content in their own function and call them inside the hook(s).
